### ISSUE TITLE
lsfd: INOTIFY.INODES column for files on btrfs

### ISSUE THREAD

**Issue by masatake** [util-linux/util-linux#2349]
@masatake FYI this fails the testsuite for me reliably:

```
--- /home/t-8ch/src/util-linux/tests/expected/lsfd/mkfds-inotify	2023-06-22 19:16:23.736001931 +0200
+++ /home/t-8ch/src/util-linux/build-meson/tests/output/lsfd/mkfds-inotify	2023-06-27 14:30:35.634742182 +0200
@@ -1,4 +1,14 @@
 INOTIFY.INODES.RAW: 0
-INOTIFY.INODES.RAW == FSTAB\x0aROOT
+INOTIFY.INODES.RAW: 1261373@0:24\x0a256@0:24
+FSTAB: 1261373@0:26
+ROOT: 256@0:26
 INOTIFY.INODES.RAW (JSON): 0
-INOTIFY.INODES.RAW == EXPECTED_JSON
+INOTIFY.INODES.RAW: {
+   "lsfd": [
+      {
+         "inotify.inodes.raw": [
+             "1261373@0:24", "256@0:24"
+         ]
+      }
+   ]
+}
}}}-diff
```

Maybe because it's on btrfs.

_Originally posted by @t-8ch in https://github.com/util-linux/util-linux/issues/2308#issuecomment-1609414473_

**Comment by masatake** [comment]
@t-8ch Thank you for reporting. I'm new to btrfs. I tried reproducing with using /dev/loop. However, it was not reproduced. Could you show the output of findmnt for the fs?
```
$ findmnt /home/yamato/var/util-linux/btrfs
TARGET                            SOURCE     FSTYPE OPTIONS
/home/yamato/var/util-linux/btrfs /dev/loop8 btrfs  rw,relatime,seclabel,discard=async,space_cache=v2,subvolid=5,subvol=/
```

**Comment by t-8ch** [comment]
```
$ findmnt /
TARGET  SOURCE             FSTYPE OPTIONS
/       /dev/mapper/system btrfs  rw,relatime,ssd,discard=async,space_cache=v2,subvolid=5,subvol=/
```

(It's using luks, the kernel is 6.3.9-arch1-1)

**Comment by t-8ch** [comment]
I can also reproduce it with a plain btrfs loop mount.

**Comment by t-8ch** [comment]
Neither xfs nor ext4 exhibit the same problem.

Different btrfs filesystems exhibit different offsets.
I have seen offsets of 1, 2 and 3.

**Comment by masatake** [comment]
offsets？ Offsets of what? Btrfs specific concept?

**Comment by t-8ch** [comment]
offset: difference in device number:

`0:24` -> `0:26`: offset 2

(made up by me)

**Comment by masatake** [comment]
I installed Fedora38 that uses Btrfs as / to a VM. The issue is reproduced.


test_mkfds used in the test case add /etc/fstab to the inotfy's monitor list.

```
[yamato@localhost-live ~]$ stat /etc/fstab 
  File: /etc/fstab
  Size: 627             Blocks: 8          IO Block: 4096   regular file
Device: 0,37    Inode: 640         Links: 1
```

```
[yamato@localhost-live util-linux]$ strace ./test_mkfds inotify 3
...
inotify_init()                          = 3
inotify_add_watch(3, "/", IN_DELETE)    = 1
inotify_add_watch(3, "/etc/fstab", IN_DELETE) = 2
...
getpid()                                = 34266
...
...
pselect6(1, [0], NULL, NULL, NULL, {sigmask=[], sigsetsize=8}
```
```
$ cat /proc/34059/fdinfo/3 
pos:    0
flags:  00
mnt_id: 15
ino:    1060
inotify wd:2 ino:280 sdev:23 mask:200 ignored_mask:0 fhandle-bytes:14 fhandle-type:4d f_handle:800200000000000000010000000000000a000000
inotify wd:1 ino:100 sdev:23 mask:200 ignored_mask:0 fhandle-bytes:14 fhandle-type:4d f_handle:0001000000000000000100000000000007000000
```

`ino:280` in `/proc/34059/fdinfo/3` matches with `Inode: 640` in the output of the stat command. `Inode` field of fdinfo uses 16 as base unlike the output of stat that uses 10 as base.

`sdev:23` in fdinfo doesn't match with `Device: 0,37` in the stat output. 
```
$ printf "%d\n" 0x23
35
```

The offset is 2.

**Comment by masatake** [comment]
I'll jump into the kernel with systemtap.

**Comment by masatake** [comment]
Both `stat` command and the `inotify` field in `fdinfo` refer to an inode.

`filename_lookup`(https://elixir.bootlin.com/linux/v6.2.9/source/fs/namei.c#L2495) is the function getting the inode for a given file name. `filename_lookup` returns a `struct path`.
Via `path->detnry->inode`, the caller of `filename_lookup` can get the inode.

`stat` command calls `statx` system call. `statx` calls 
`filename_lookup` eventually.

`inotify_add_watch` system call takes a file name. The `inotify_add_watch` calls the `filename_lookup` eventually for getting the inode for the file name.
The inode number that `inotify_add_watch` gets via `filename_lookup` is printed in the `inotify` field in `fdinfo`.

The device number, the subject of this issue, can be obtained via `path->detnry->inode->i_sb->s_dev`. Both the `stat` command and
the `inotify` field in `fdinfo` use the `filename_lookup` for getting `path`.  If they use the same function, why don't the device numbers match? I monitored the device numbers obtained via `path->detnry->inode->i_sb->s_dev` by inserting a systemtap probe to `filename_lookup`. I saw the numbers matched.

However, the number monitored via systemtap did not match the number printed by the `stat` command.
`statx` system call doesn't use `path->detnry->inode->i_sb->s_dev` , the value obtained via `filename_lookup`, directly.
`statx` calls `vfs_statx`. `vfs_statx` calls `vfs_getattr` after calling the `filename_lookup` for filling `struct kstat`. `vfs_getattr` calls `inode->i_op->getattr`, a file system specific method for filling `struct kstat` if it is available.
btrfs has an implementation for the method, `btrfs_getattr`(https://elixir.bootlin.com/linux/v6.2.9/source/fs/btrfs/inode.c#L9007):

|  [stat](https://elixir.bootlin.com/linux/v6.2.9/C/ident/stat)->dev = [BTRFS_I](https://elixir.bootlin.com/linux/v6.2.9/C/ident/BTRFS_I)([inode](https://elixir.bootlin.com/linux/v6.2.9/C/ident/inode))->root->[anon_dev](https://elixir.bootlin.com/linux/v6.2.9/C/ident/anon_dev);

The dev member is overwritten with btrfs specific value.

Now I understand what happens behind the scene.
I got more questions. 
<1> How about other file systems?
<2> How do the other commands in util-linux handle the device number mismatching?
<3> What should we do next?

**Comment by masatake** [comment]
After studying I found https://unix.stackexchange.com/questions/345220/btrfs-how-to-get-real-device-id . I cannot find this page before this self-studying.

**Comment by masatake** [comment]
> adjusting the output in lsfd the algorithm: ...

I'll take this approach. The algorithm may be reusable in lslocks.

**Comment by karelzak** [comment]
It reminds my 13 years of discussion about stat() with btrfs devels https://bugzilla.redhat.com/show_bug.cgi?id=711881 :-)

It's the reason why libmount uses paths (e.e. mnt_table_find_target()) from the beginning rather than device numbers as usual on UNIXes.

**Comment by masatake** [comment]
After rethinking this issue, I recognized this is not a bug of lsfd but a bug of the test case (tests/ts/lsfd/mkfsd-inotify).
I made a pull request to fix the test case (#2683).

>> Adjusting the output in lsfd the algorithm: ...

> I'll take this approach. The algorithm may be reusable in lslocks.

The algorithm itself is useful, but implementing the algorithm doesn't solve the issue.

During the whole holiday week, I have implemented this one.
Though the algorithm doesn't solve this issue, the algorithm improves lsfd in at least two aspects:

* decoding the SOURCE column for a fd opening a file on btrfs

  Without the algorithm:

  ```
  $ ./lsfd -Q '(ASSOC == "exe")' -p $$
  COMMAND   PID   USER ASSOC  XMODE TYPE SOURCE MNTID  INODE NAME
  zsh     19318 yamato   exe ------  REG   0:38     0 589767 /usr/bin/zsh
  ```
  With the algorithm:
  ```
  $ ./lsfd -Q '(ASSOC == "exe")' -p $$
  COMMAND   PID   USER ASSOC  XMODE TYPE SOURCE MNTID  INODE NAME
  zsh     19318 yamato   exe ------  REG  btrfs     0 589767 /usr/bin/zsh
  ```

  SOURCE field can be decoded.

* #2567
